### PR TITLE
feat(math): add baby and koala bear fields

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR adds baby bear and koala bear prime fields.

### CODE REVIEW

File: tachyon/math/finite_fields/packed_prime_field32_avx2.h [hunk 1]
```diff
@@ -19,28 +19,38 @@ ALWAYS_INLINE __m256i AddMod32(__m256i lhs, __m256i rhs, __m256i p) {
...
   //
-  // 1) if 0 ≤ t ≤ p - 1:
-  //    2³² - p ≤ u ≤ 2³² - 1
-  //    2(p + 1) - p ≤ u ≤ 2³² - 1
-  //    p - 1 < p + 1 ≤ u ≤ 2³² - 1
+  // 1) lhs = p && rhs ≠ p
+  //    0 ≤ t ≤ p - 1, go to 4)
+  //
+  // 2) lhs ≠ p && rhs = p
+  //    0 ≤ t ≤ p - 1, go to 4)
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx2.h:31]: Do you mean 5 here?

File: tachyon/math/finite_fields/packed_prime_field32_avx512.h [hunk 2]
```diff
@@ -19,28 +19,38 @@ ALWAYS_INLINE __m512i AddMod32(__m512i lhs, __m512i rhs, __m512i p) {
...
   //
-  // 1) if 0 ≤ t ≤ p - 1:
-  //    2³² - p ≤ u ≤ 2³² - 1
-  //    2(p + 1) - p ≤ u ≤ 2³² - 1
-  //    p - 1 < p + 1 ≤ u ≤ 2³² - 1
+  // 1) lhs = p && rhs ≠ p
+  //    0 ≤ t ≤ p - 1, go to 4)
+  //
+  // 2) lhs ≠ p && rhs = p
+  //    0 ≤ t ≤ p - 1, go to 4)
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx512.h:31]: Do you mean 5 here?

File: tachyon/math/finite_fields/packed_prime_field32_avx2.h [hunk 3]
```diff
@@ -19,28 +19,38 @@ ALWAYS_INLINE __m256i AddMod32(__m256i lhs, __m256i rhs, __m256i p) {
...
   //    r = t
   //
-  // 2) otherwise p ≤ t ≤ 2p - 2:
-  //    0 ≤ u ≤ p - 2 < p
+  // 5) p ≤ t ≤ 2p - 2
+  //    0 ≤ u ≤ p - 2 ≤ p
   //    r = u
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx2.h:46]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_avx2.h [hunk 4]
```diff
@@ -52,23 +62,33 @@ ALWAYS_INLINE __m256i SubMod32(__m256i lhs, __m256i rhs, __m256i p) {
...
+  // 5) -p + 1 ≤ d ≤ -1
   //    2³² - p + 1 ≤ t ≤ 2³² - 1
-  //    2(p + 1) - p + 1 ≤ t ≤ 2³² - 1
-  //    p + 3 ≤ t ≤ 2³² - 1
-  //    1 ≤ u ≤ p - 1 < p + 3
+  //    1 ≤ u ≤ p - 1
   //    r = u
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx2.h:91]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_avx2.h [hunk 5]
```diff
@@ -83,21 +103,27 @@ ALWAYS_INLINE __m256i NegMod32(__m256i val, __m256i p) {
...
+  //    r = t
+  //    t = 0
+  //
+  // 3) 1 ≤ val ≤ p - 1
   //    r = t
   //    2³² - p + 1 ≤ -val ≤ 2³² - 1
   //    1 ≤ t ≤ p - 1
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx2.h:126]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_avx512.h [hunk 6]
```diff
@@ -19,28 +19,38 @@ ALWAYS_INLINE __m512i AddMod32(__m512i lhs, __m512i rhs, __m512i p) {
...
   //    r = t
   //
-  // 2) otherwise p ≤ t ≤ 2p - 2:
-  //    0 ≤ u ≤ p - 2 < p
+  // 5) p ≤ t ≤ 2p - 2
+  //    0 ≤ u ≤ p - 2 ≤ p
   //    r = u
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx512.h:46]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_avx512.h [hunk 7]
```diff
@@ -52,23 +62,33 @@ ALWAYS_INLINE __m512i SubMod32(__m512i lhs, __m512i rhs, __m512i p) {
...
+  // 5) -p + 1 ≤ d ≤ -1
   //    2³² - p + 1 ≤ t ≤ 2³² - 1
-  //    2(p + 1) - p + 1 ≤ t ≤ 2³² - 1
-  //    p + 3 ≤ t ≤ 2³² - 1
-  //    1 ≤ u ≤ p - 1 < p + 3
+  //    1 ≤ u ≤ p - 1
   //    r = u
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_avx512.h:91]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_neon.h [hunk 8]
```diff
@@ -41,28 +40,38 @@ ALWAYS_INLINE uint32x4_t AddMod32(uint32x4_t lhs, uint32x4_t rhs,
...
+  //
+  // 0 ≤ lhs, rhs ≤ m
+  //
+  // 1) lhs = p && rhs ≠ p
+  //    0 ≤ t ≤ p - 1, go to 4)
   //
-  // 0 ≤ lhs, rhs ≤ p - 1
-  // 0 ≤ t ≤ 2p - 2
+  // 2) lhs ≠ p && rhs = p
+  //    0 ≤ t ≤ p - 1, go to 4)
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_neon.h:52]: Do you mean 5 here?

File: tachyon/math/finite_fields/packed_prime_field32_neon.h [hunk 9]
```diff
@@ -41,28 +40,38 @@ ALWAYS_INLINE uint32x4_t AddMod32(uint32x4_t lhs, uint32x4_t rhs,
...
   //    r = t
   //
-  // 2) otherwise p ≤ t ≤ 2p - 2:
-  //    0 ≤ u ≤ p - 2 < p
+  // 5) p ≤ t ≤ 2p - 2
+  //    0 ≤ u ≤ p - 2 ≤ p
   //    r = u
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_neon.h:67]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_neon.h [hunk 10]
```diff
@@ -108,14 +129,27 @@ ALWAYS_INLINE uint32x4_t NegMod32(uint32x4_t val, uint32x4_t p) {
...
+  //    t = 0
+  //    r = 0
+  //
+  // 3) 1 ≤ val ≤ p - 1
+  //    is_zero = 0
+  //    2³² - p + 1 ≤ -val ≤ 2³² - 1
+  //    1 ≤ t ≤ p - 1
   //    r = t
+  //
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_neon.h:152]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_neon.h [hunk 11]
```diff
@@ -74,21 +83,33 @@ ALWAYS_INLINE uint32x4_t SubMod32(uint32x4_t lhs, uint32x4_t rhs,
...
-  //    2³² -p + 1 - (2³² - 1)p ≤ r ≤ 2³² - 1 - (2³² - 1)p
-  //    2³²(1 - p) + 1 ≤ r ≤ 2³²(1 - p) + p - 1
-  //    1 ≤ r ≤ p - 1
+  // 5) -p + 1 ≤ d ≤ -1
+  //    2³² - p + 1 ≤ t ≤ 2³² - 1
+  //    1 ≤ u ≤ p - 1
+  //    r = u
   //
-  // In both cases, r is in {0, ..., p - 1}.
+  // In all the cases, r is in {0, ..., m}.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_neon.h:112]:
```suggestion
  // In all cases, r is in {0, ..., m}.
```

File: tachyon/math/finite_fields/packed_prime_field32_neon.h [hunk 12]
```diff
@@ -74,21 +83,33 @@ ALWAYS_INLINE uint32x4_t SubMod32(uint32x4_t lhs, uint32x4_t rhs,
...
+  //          { p - 1 otherwise
+  //
+  // 0 ≤ lhs, rhs ≤ m
+  //
+  // 1) lhs = p && rhs ≠ p
+  //    1 ≤ d ≤ p, go to 4)
   //
-  // 0 ≤ lhs, rhs ≤ p - 1
+  // 2) lhs ≠ p && rhs = p
+  //    -p ≤ d ≤ -1, go to 5)
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/packed_prime_field32_neon.h:95]:
```suggestion
  //    -p + 1 ≤ d ≤ -1, go to 5)
```

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc [hunk 13]
```diff
@@ -62,44 +62,96 @@ __m256i Mul(__m256i lhs, __m256i rhs) {
...
+  // clang-format off
+  //                  ┌────────────────┬─────────────────────────────┬─────┬─────────────────┬─────────────────────────────┐
+  // |prod_lo_dirty|: │ Lo(lhs₀ * rhs₀)│ (lhs₁ * rhs₁)[0..31](dirty) │ ... │ Lo(lhs₆ * rhs₆) │ (lhs₇ * rhs₇)[0..31](dirty) │
+  //                  └────────────────┴─────────────────────────────┴─────┴─────────────────┴─────────────────────────────┘
+  // clang-format on
   __m256i prod_lo_dirty =
       _mm256_blend_epi32(prod_evn, prod_odd_lo_dirty, 0b10101010);
   // Now put all the high halves into one vector. The even values come from
-  // prod_evn_hi and the odd values come from prod_odd_dbl.
+  // prod_evn_hi and the odd values come from |prod_odd_dbl|.
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc:133]:
```suggestion
  // |prod_evn_hi| and the odd values come from |prod_odd_dbl|.
```

File: tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl [hunk 14]
```diff
@@ -167,16 +168,22 @@ def _do_generate_prime_fields(
             "//tachyon:export",
             "//tachyon/build:build_config",
             "//tachyon/math/base:big_int",
+            "@com_google_absl//absl/numeric:int128",
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl:171]: Why is this added?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/math/finite_fields/packed_prime_field32_avx2.h (hunk 1 above)
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field32_avx2.h:31]: I think 4 is correct.

File: tachyon/math/finite_fields/packed_prime_field32_avx512.h (hunk 2 above)
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field32_avx512.h:31]: I think 4 is correct.

File: tachyon/math/finite_fields/packed_prime_field32_neon.h (hunk 8 above)
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field32_neon.h:52]: 4 is correct.

File: tachyon/math/finite_fields/packed_prime_field32_neon.h (hunk 12 above)
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field32_neon.h:95]: I meant that range is -p ≤ d ≤ -1 and -p + 1 ≤ d ≤ -1, which is case 5, includes the range so it has to go to case 5.

File: tachyon/math/finite_fields/small_prime_field_mont.h [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+                                           _Config::kUseMontgomery>>
+    final : public PrimeFieldBase<PrimeField<_Config>> {
+ public:
+  constexpr static size_t kModulusBits = _Config::kModulusBits;
+  constexpr static size_t kLimbNums = (kModulusBits + 63) / 64;
+  constexpr static size_t N = kLimbNums;
+
+  using Config = _Config;
+  using BigIntTy = BigInt<N>;
+  using MontgomeryTy = BigInt<N>;
```
Comment (Insun35) [inline tachyon/math/finite_fields/small_prime_field_mont.h:42]: Are these two aliases for same type for readability? How about using `BigIntTy` for both cases?

File: tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl (hunk 14 above)
Comment (chokobole) [reply tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl:171]: Oh this is an intermediate result, now it should be removed ,Thanks!

File: tachyon/math/finite_fields/baby_bear/packed_baby_bear_avx2.cc [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  _mm256_storeu_si256(reinterpret_cast<__m256i_u*>(ret.values().data()),
+                      vector);
+  return ret;
+}
+
+__m256i Add(__m256i lhs, __m256i rhs) { return AddMod32(lhs, rhs, kP); }
+
+__m256i Sub(__m256i lhs, __m256i rhs) { return SubMod32(lhs, rhs, kP); }
+
+__m256i Neg(__m256i val) { return NegMod32(val, kP); }
```
Comment (Insun35) [inline tachyon/math/finite_fields/baby_bear/packed_baby_bear_avx2.cc:37]: I thought we decided to use full name for unary operations. If so, please change it to `Negate`

Comment (chokobole) [reply tachyon/math/finite_fields/baby_bear/packed_baby_bear_avx2.cc:37]: Oh got it, that makes sense!

File: tachyon/math/finite_fields/small_prime_field_mont.h (hunk 15 above)
Comment (chokobole) [reply tachyon/math/finite_fields/small_prime_field_mont.h:42]: Yeap, I agree that the naming convention I used initially was not optimal. The `MontgomeryTy` type is intended for facilitating data transfer between various prime field implementations, such as `PrimeFieldGpu` and `PrimeField`. You can find numerous examples by searching for `ToMontgomery()`. I believe it would be beneficial to rename these to convey a clearer message

**Review by Insun35 — APPROVED** [review]
LGTM

**Comment by dongchangYoo** [comment]
In commit body of https://github.com/kroma-network/tachyon/pull/405/commits/8df7697cbde1ca5be4357d90a262a4f5267b4732, typo exists. Please fix `whre` to `where`

File: tachyon/math/finite_fields/prime_field_unittest.cc [hunk 17]
```diff
@@ -1,7 +1,7 @@
 #include "gtest/gtest.h"
 
 #include "tachyon/base/buffer/vector_buffer.h"
-#include "tachyon/math/finite_fields/prime_field_gpu_debug.h"
+// #include "tachyon/math/finite_fields/prime_field_gpu_debug.h"
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/prime_field_unittest.cc:4]: This is needed, please read the comment below in the same file

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc [hunk 18]
```diff
@@ -54,7 +54,7 @@ __m256i Mul(__m256i lhs, __m256i rhs) {
   // vpblendd   prod_hi, prod_evn_hi, prod_odd_dbl, aah
   // vpand      prod_lo, prod_lo_dirty, P
```
Comment (ashjeong) [inline tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc:55]:
```suggestion
  // vpand      prod_lo, prod_lo_dirty, p

**Review by ashjeong — COMMENTED** [review]
Please fix the commit title and body as such:

chore(math): update comments to contain montgomery form **& fix minor errors**
**For** the 31-bit prime field, the montgomery form is `a * R`, **where** R is 2³² **, so** value `a` can be **in the range** {0, ..., **p**}, **where** **p** is **the** modulus. The
comments are **also** updated to handle this case.

Additionally, this commit also **contains** minor fixes, such as **changing** `P`
to `p` and removing **commas** and **periods**.

File: tachyon/math/finite_fields/packed_prime_field32_avx2.h [hunk 19]
```diff
@@ -19,28 +19,40 @@ ALWAYS_INLINE __m256i AddMod32(__m256i lhs, __m256i rhs, __m256i p) {
...
+  //    u = 0
+  //    r = 0
+  //
+  // 2) lhs = p && rhs = p
+  //    t = 2p
+  //    u = p
+  //    r = p, which is 0.
+  //
+  // 3) (lhs = p && 1 ≤ rhs ≤ p - 1) || (1 ≤ lhs ≤ p - 1 && rhs = p)
+  //    p + 1 ≤ t ≤ 2p - 1, go to 5)
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/packed_prime_field32_avx2.h:38]: But when you see the each item, it starts with e.g, `x)` which is very normal to me. Then, do you think it's better to change numbering `x)` to `(x)`?

[the remark this answers is not in the document]
Comment (ashjeong) [reply tachyon/math/finite_fields/packed_prime_field32_avx2.h:38]: It seems to boil down to preference :thinking:. Looking at it more closely, can the branches just be merged? (aka in this case logic branch 3 & 5 would be merged). It doesn't seem like the x branches in "Go to x" are redundant.

ex:
```diff
  // 1) (lhs = p && rhs = 0) || (lhs = 0 && rhs = p)
  //    t = p
  //    u = 0
  //    r = 0
  //
  // 2) lhs = p && rhs = p
  //    t = 2p
  //    u = p
  //    r = p, which is 0.
  //
  // 3) (lhs = p && 1 ≤ rhs ≤ p - 1) || (1 ≤ lhs ≤ p - 1 && rhs = p)
-  //    p + 1 ≤ t ≤ 2p - 1, go to 5)
+  //    p + 1 ≤ t ≤ 2p - 1
+  //    1 ≤ u ≤ p - 1 ≤ p
+  //    r = u
  //
  // 4) 0 ≤ t ≤ p - 1
  //    p < 2³² - p ≤ u ≤ 2³² - 1
  //    r = t
  //
  // 5) p + 1 ≤ t ≤ 2p - 1
-  //    1 ≤ u ≤ p - 1 ≤ p
-  //    r = u
  ```

[the remark this answers is not in the document]
Comment (ashjeong) [reply tachyon/math/finite_fields/packed_prime_field32_avx2.h:38]: It seems to boil down to preference :thinking:. To be fair, once I continued to read the code, I understood the notation as well. I suppose in this case, the change it trivial, so I'll resolve this.

File: tachyon/math/finite_fields/packed_prime_field32_avx512.h [hunk 20]
```diff
@@ -126,6 +126,145 @@ ALWAYS_INLINE __m512i NegateMod32(__m512i val, __m512i p) {
...
+  //      vpsubd     r, prod_hi, q_p_hi
+  //      vpaddd     r{underflow}, r, p
+  // throughput: 6.5 cyc/vec (2.46 els/cyc)
+  // latency: 21 cyc
+
+  // |vpmuludq| only reads the even doublewords, so when we pass |lhs| and
+  // |rhs| directly we get the eight products at even positions.
+  //            ┌──────┬──────┬─────┬───────┬───────┐
+  // |lhs_evn|: │ lhs₀ │ lhs₁ │ ... │ lhs₁₄ │ lhs₁₅ │
+  //            └───┴──────┴─────┴───────┴───────┘
```
Comment (ashjeong) [inline tachyon/math/finite_fields/packed_prime_field32_avx512.h:184]:
```suggestion
  // |lhs_evn|: │ lhs₀ │ lhs₁ │ ... │ lhs₁₄ │ lhs₁₅ │
  //            └──────┴──────┴─────┴───────┴───────┘

**Review by ashjeong — CHANGES_REQUESTED** [review]
https://github.com/kroma-network/tachyon/pull/405/commits/def41fcb0d981fe6aa63142bc8dedcb2ccb0a3c5
refac(math): rename NegXXX to NegateXXX for **consistency**

https://github.com/kroma-network/tachyon/pull/405/commits/00e1b88ba7519c6eb80413f235dbe87186b11857
chore(math): add comments for **`PackedMersenne31AVX`** **multiplication**

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM
